### PR TITLE
Respect HTTP client options when reading remote requirements files

### PR DESCRIPTION
**Pull request by zanieb**
Uses the base client introduced in #2431 to restore use of our fully configured client when reading remote requirements files.

Closes https://github.com/astral-sh/uv/issues/2357

## Test plan

```
npx http-server --username user --password password
cargo run -- pip install -r http://user:password@127.0.0.1:8080/requirements.txt
```

Fails on main succeeds on branch.

### CODE REVIEW

File: crates/uv/src/requirements.rs [hunk 1]
```diff
@@ -281,15 +281,15 @@ impl RequirementsSpecification {
...
+        client_builder: BaseClientBuilder,
     ) -> Result<Self> {
         let mut spec = Self::default();
 
         // Read all requirements, and keep track of all requirements _and_ constraints.
         // A `requirements.txt` can contain a `-c constraints.txt` directive within it, so reading
         // a requirements file can also add constraints.
         for source in requirements {
-            let source = Self::from_source(source, extras, connectivity).await?;
+            let source = Self::from_source(source, extras, client_builder.clone()).await?;
```
Comment (zanieb) [inline crates/uv/src/requirements.rs:292]: These clones are alarming to me. We should pass a reference to a lazily built client instead?

Comment (zanieb) [reply crates/uv/src/requirements.rs]: Addressed here in https://github.com/astral-sh/uv/pull/2434/commits/2bcf617d6a4fe43538e2ceb4209178542495b726 instead

**Comment by zanieb** [comment]
Should add a test case for this

File: crates/uv-client/src/base_client.rs [hunk 2]
```diff
@@ -88,7 +94,11 @@ impl<'a> BaseClientBuilder<'a> {
         self
     }
 
-    pub fn build(self) -> BaseClient {
+    pub fn is_offline(&self) -> bool {
+        matches!(self.connectivity, Connectivity::Offline)
+    }
+
+    pub fn build(&self) -> BaseClient {
```
Comment (charliermarsh) [inline crates/uv-client/src/base_client.rs:101]: In lieu of calling `build` multiple times, did you explore trying to do this with interior mutability? E.g., making `client()` a `OnceCell` or something lazily constructed?

Comment (zanieb) [reply crates/uv-client/src/base_client.rs:101]: Ah I looked at this a little in #2439 — I'd be happy to explore that more next.

**Comment by zanieb** [comment]
@charliermarsh did you feel like we need lazy building before merging this?

**Review by charliermarsh — APPROVED** [review]
Na, it's an improvement for users and in correctness.

**Pull request merged**